src/probability/martingale/optional_sampling.lean

```diff
@@ -0,0 +1,266 @@
+begin
+  haveI : sigma_finite (μ.trim hτ.measurable_space_le),
+  { exact sigma_finite_trim_mono _ (is_stopping_time.measurable_space_mono hσ hτ hσ_le_τ), },
+  exact h.stopped_value_ae_eq_condexp_of_le_of_countable_range hτ hσ hσ_le_τ hτ_le
+    (set.to_countable _) (set.to_countable _),
+end
+
+end first_countable_topology
+
+section subset_of_nat
+
+/-! In the following results the index set verifies `[linear_order ι] [locally_finite_order ι]` and
+`[order_bot ι]`, which means that it is order-isomorphic to a subset of `ℕ`. `ι` is equipped with
+the discrete topology, which is also the order topology, and is a measurable space with the Borel
+σ-algebra. -/
+
+variables {ι : Type*} [linear_order ι] [locally_finite_order ι] [order_bot ι]
+  [topological_space ι] [discrete_topology ι] [measurable_space ι] [borel_space ι]
+  [measurable_space E] [borel_space E] [second_countable_topology E]
+  {ℱ : filtration ι m} {τ σ : Ω → ι} {f : ι → Ω → E} {i n : ι}
+
+lemma condexp_stopped_value_stopping_time_ae_eq_restrict_le
+  (h : martingale f ℱ μ) (hf_prog : prog_measurable ℱ f)
+  (hτ : is_stopping_time ℱ τ) (hσ : is_stopping_time ℱ σ)
+  [sigma_finite (μ.trim hσ.measurable_space_le)] (hτ_le : ∀ x, τ x ≤ n) :
+  μ[stopped_value f τ | hσ.measurable_space] =ᵐ[μ.restrict {x : Ω | τ x ≤ σ x}] stopped_value f τ :=
+begin
+  rw ae_eq_restrict_iff_indicator_ae_eq
+    (hτ.measurable_space_le _ (hτ.measurable_set_le_stopping_time hσ)),
+  swap, apply_instance,
+  refine (condexp_indicator (integrable_stopped_value ι hτ h.integrable hτ_le)
+    (hτ.measurable_set_stopping_time_le hσ)).symm.trans _,
+  have h_int : integrable ({ω : Ω | τ ω ≤ σ ω}.indicator (stopped_value (λ (n : ι), f n) τ)) μ,
+  { refine (integrable_stopped_value ι hτ h.integrable hτ_le).indicator _,
+    exact hτ.measurable_space_le _ (hτ.measurable_set_le_stopping_time hσ), },
+  have h_meas : ae_strongly_measurable' hσ.measurable_space
+    ({ω : Ω | τ ω ≤ σ ω}.indicator (stopped_value (λ (n : ι), f n) τ)) μ,
+  { refine strongly_measurable.ae_strongly_measurable' _,
+    refine strongly_measurable.strongly_measurable_of_measurable_space_le_on
+      (hτ.measurable_set_le_stopping_time hσ) _ _ _,
+    { intros t ht,
+      rw set.inter_comm _ t at ht ⊢,
+      rw [hτ.measurable_set_inter_le_iff, is_stopping_time.measurable_set_min_iff hτ hσ] at ht,
+      exact ht.2, },
+    { refine strongly_measurable.indicator _ (hτ.measurable_set_le_stopping_time hσ),
+      refine measurable.strongly_measurable _,
+      exact measurable_stopped_value hf_prog hτ, },
+    { intros x hx,
+      simp only [hx, set.indicator_of_not_mem, not_false_iff], }, },
+  exact condexp_of_ae_strongly_measurable' hσ.measurable_space_le h_meas h_int,
+end
+
+/-- **Optional Sampling theorem**. If `τ` is a bounded stopping time and `σ` is another stopping
+time, then the stopped value of a martingale `f` with respect to `min τ σ` is almost everywhere
+equal to `μ[stopped_value f τ | hσ.measurable_space]`. -/
```

Comment:
could you use a sentence instead of a formula to explain the theorem? Same thing in the next theorem.